File: app/src/main/java/com/antonio/samir/meteoritelandingsspots/ui/recyclerView/MeteoriteAdapter.java
package com.antonio.samir.meteoritelandingsspots.ui.recyclerView;

import android.content.ContentResolver;
import android.content.Context;
import android.database.ContentObserver;
import android.database.Cursor;
import android.net.Uri;
import android.os.Handler;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.antonio.samir.meteoritelandingsspots.R;
import com.antonio.samir.meteoritelandingsspots.service.repository.MeteoriteProvider;
import com.antonio.samir.meteoritelandingsspots.service.server.AddressService;
import com.antonio.samir.meteoritelandingsspots.ui.recyclerView.selector.MeteoriteSelector;

import org.apache.commons.lang3.StringUtils;

import static com.antonio.samir.meteoritelandingsspots.service.repository.MeteoriteColumns.ID;
import static com.antonio.samir.meteoritelandingsspots.service.repository.MeteoriteColumns.NAME;
import static com.antonio.samir.meteoritelandingsspots.service.repository.MeteoriteColumns.YEAR;

/**
 * Custom RecyclerView.Adapter to deal with meteorites cursor
 */
public class MeteoriteAdapter extends CursorRecyclerViewAdapter<ViewHolderMeteorite> {
    private final MeteoriteSelector meteoriteSelector;
    private Context mContext;
    private ContentObserver observer;

    public MeteoriteAdapter(Context context, Cursor cursor, final MeteoriteSelector meteoriteSelector) {
        super(context, cursor);
        this.mContext = context;
        this.meteoriteSelector = meteoriteSelector;
    }

    @Override
    public ViewHolderMeteorite onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_meteorite, parent, false);
        final ViewHolderMeteorite vh = new ViewHolderMeteorite(view);

        //On view click use MeteoriteSelector to do execute the proper according the current layout
        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                meteoriteSelector.selectItemId(vh.getId());
            }
        });
        return vh;
    }

    @Override
    public int getItemCount() {
        final int itemCount = super.getItemCount();
        return itemCount;
    }

    public void resetData() {
        swapCursor(null);
    }

    public void setData(Cursor data) {
        swapCursor(data);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public void onBindViewHolder(final ViewHolderMeteorite viewHolder, final Cursor cursor) {

        final String meteoriteName = cursor.getString(cursor.getColumnIndex(NAME));
        final String year = cursor.getString(cursor.getColumnIndex(YEAR));

        final int id = cursor.getColumnIndex(ID);
        final String idString = cursor.getString(id);

        viewHolder.name.setText(meteoriteName);

        viewHolder.year.setText(year);

        recoverAddress(viewHolder, idString);

        viewHolder.setId(idString);
    }

    private void recoverAddress(final ViewHolderMeteorite viewHolder, final String idString) {

        final ContentResolver contentResolver = mContext.getContentResolver();

        final AddressService addressService = new AddressService(contentResolver);

        String address = addressService.getAddressFromId(idString);

        if (StringUtils.isBlank(address)) {

            final Uri uri = MeteoriteProvider.Addresses.withId(idString);

            observer = new ContentObserver(new Handler()) {
                @Override
                public void onChange(boolean selfChange) {
                    super.onChange(selfChange);
                    final String address = addressService.getAddressFromId(idString);
                    viewHolder.location.setText(address);
                    //mContext.getContentResolver().unregisterContentObserver(observer);
                }
            };

            contentResolver.registerContentObserver(uri, true, observer);

        } else {
            viewHolder.location.setText(address);
        }


    }

    @Override
    public void onDetachedFromRecyclerView(RecyclerView recyclerView) {
        super.onDetachedFromRecyclerView(recyclerView);
        if (observer != null) {
            mContext.getContentResolver().unregisterContentObserver(observer);
        }
    }

}

File: app/src/main/java/com/antonio/samir/meteoritelandingsspots/ui/recyclerView/MeteoriteAdapter.kt
package com.antonio.samir.meteoritelandingsspots.ui.recyclerView

import androidx.lifecycle.Observer
import android.content.Context
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils

import com.antonio.samir.meteoritelandingsspots.R
import com.antonio.samir.meteoritelandingsspots.model.Meteorite
import com.antonio.samir.meteoritelandingsspots.presenter.MeteoriteListPresenter
import com.antonio.samir.meteoritelandingsspots.ui.recyclerView.selector.MeteoriteSelector

import org.apache.commons.lang3.StringUtils

/**
 * Custom RecyclerView.Adapter to deal with meteorites cursor
 */
class MeteoriteAdapter(private val mContext: Context, private val meteoriteSelector: MeteoriteSelector, private val mPresenter: MeteoriteListPresenter) : androidx.recyclerview.widget.RecyclerView.Adapter<ViewHolderMeteorite>() {
    private var mSelectedMeteorite: String? = null
    var vieHolderMeteorite: ViewHolderMeteorite? = null
        private set
    private var mMeteorites: List<Meteorite>? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolderMeteorite {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.list_item_meteorite, parent, false)
        val vh = ViewHolderMeteorite(view)

        //On view click use MeteoriteSelector to do execute the proper according the current layout
        view.setOnClickListener { view1 ->
            vieHolderMeteorite = vh
            meteoriteSelector.selectItemId(vh.id)
        }
        return vh
    }


    override fun getItemCount(): Int {
        return if (mMeteorites != null) mMeteorites!!.size else 0
    }

    fun setData(data: List<Meteorite>) {
        mMeteorites = data
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun onBindViewHolder(viewHolder: ViewHolderMeteorite, position: Int) {
        val meteorite = mMeteorites!![position]

        val meteoriteName = meteorite.name
        val year = meteorite.yearString

        val idString = meteorite.id.toString()

        viewHolder.mName.text = mContext.getString(R.string.title, meteoriteName, year)
        viewHolder.mYear.text = year
        viewHolder.mYear.visibility = View.GONE

        viewHolder.mName.contentDescription = meteoriteName
        viewHolder.mYear.contentDescription = year

        setLocationText(meteorite, viewHolder)

        viewHolder.id = idString


        var color = R.color.unselected_item_color
        var title_color = R.color.title_color
        var elevation = R.dimen.unselected_item_elevation

        if (StringUtils.equals(idString, mSelectedMeteorite)) {
            color = R.color.selected_item_color
            title_color = R.color.selected_title_color
            elevation = R.dimen.selected_item_elevation
        }

        viewHolder.mCardview.setCardBackgroundColor(mContext.resources.getColor(color))
        viewHolder.mCardview.cardElevation = mContext.resources.getDimensionPixelSize(elevation).toFloat()
        viewHolder.mName.setTextColor(mContext.resources.getColor(title_color))

    }


    fun setLocationText(meteorite: Meteorite, viewHolder: ViewHolderMeteorite) {
        val address = meteorite.address

        //Always remove the previous observer
        if (viewHolder.liveMet != null && viewHolder.addressObserver != null) {
            viewHolder.liveMet.removeObserver(viewHolder.addressObserver)
            viewHolder.liveMet = null
            viewHolder.addressObserver = null
        }

        if (StringUtils.isNotEmpty(address)) {
            showAddress(viewHolder, address)
        } else {
            //If address is still empty then observe this entity to be aware of any change
            viewHolder.liveMet = mPresenter.getMeteorite(meteorite)

            var addressObserver = Observer<Meteorite> { meteorite1 ->
                val newAddress = meteorite1?.getAddress()
                if (StringUtils.isNotEmpty(newAddress)) {
                    showAddress(viewHolder, newAddress!!)
                    viewHolder.mLocation.startAnimation(AnimationUtils.loadAnimation(mContext, R.anim.view_show))
                    viewHolder.liveMet.removeObserver(viewHolder.addressObserver)
                }
            }

            viewHolder.addressObserver = addressObserver
            viewHolder.liveMet.observeForever(viewHolder.addressObserver)
            viewHolder.mLocation.visibility = View.INVISIBLE
        }
    }

    private fun showAddress(viewHolder: ViewHolderMeteorite, address: String) {
        viewHolder.mLocation.text = address
        viewHolder.mLocation.visibility = View.VISIBLE
    }

    fun setSelectedMeteorite(selectedMeteorite: String) {
        notifyDataSetChanged()
        this.mSelectedMeteorite = selectedMeteorite
    }
}